File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/SQLQueryBuilder.py
class SQLQueryBuilder:

    @staticmethod
    def select(table, columns=None, where=None):
        if columns is None:
            columns = ["*"]
        elif isinstance(columns, str):
            columns = [col.strip() for col in columns.split(",")]
        return SQLQueryBuilder._select(table, columns, where)

    @staticmethod
    def _select(table, columns, where=None):
        query = "SELECT "
        if columns and columns[0] != "*":
            query += ", ".join(columns)
        else:
            query += "*"
        query += f" FROM {table}"

        if where is not None and len(where) > 0:
            query += " WHERE "
            first = True
            for key, value in where.items():
                if not first:
                    query += " AND "
                query += f"{key}='{value}'"
                first = False
        return query

    @staticmethod
    def insert(table, data):
        query = f"INSERT INTO {table} ("
        values = " VALUES ("
        first = True
        for key, value in data.items():
            if not first:
                query += ", "
                values += ", "
            query += key
            values += f"'{value}'"
            first = False
        query += ")"
        query += values + ")"
        return query

    @staticmethod
    def delete(table, where=None):
        query = f"DELETE FROM {table}"
        if where is not None and len(where) > 0:
            query += " WHERE "
            first = True
            for key, value in where.items():
                if not first:
                    query += " AND "
                query += f"{key}='{value}'"
                first = False
        return query

    @staticmethod
    def update(table, data, where=None):
        query = f"UPDATE {table} SET "
        first = True
        for key, value in data.items():
            if not first:
                query += ", "
            query += f"{key}='{value}'"
            first = False
        if where is not None and len(where) > 0:
            query += " WHERE "
            first = True
            for key, value in where.items():
                if not first:
                    query += " AND "
                query += f"{key}='{value}'"
                first = False
        return query

File: data/derived_cot/rq1_traces/translation/r1/py_to_cpp/code_output/SQLQueryBuilder.cpp
#include <string>
#include <vector>
#include <map>

class SQLQueryBuilder {
public:
    static std::string select(const std::string& table, 
                              const std::vector<std::string>& columns = {"*"}, 
                              const std::map<std::string, std::string>& where = {}) {
        std::string columns_str;
        if (columns.empty()) {
            columns_str = "*";
        } else {
            for (auto it = columns.begin(); it != columns.end(); ++it) {
                if (it != columns.begin()) {
                    columns_str += ", ";
                }
                columns_str += *it;
            }
        }

        std::string query = "SELECT " + columns_str + " FROM " + table;
        if (!where.empty()) {
            query += " WHERE ";
            for (auto it = where.begin(); it != where.end(); ++it) {
                if (it != where.begin()) {
                    query += " AND ";
                }
                query += it->first + "='" + it->second + "'";
            }
        }
        return query;
    }

    static std::string insert(const std::string& table, 
                              const std::map<std::string, std::string>& data) {
        if (data.empty()) {
            return "INSERT INTO " + table + " () VALUES ()";
        }
        std::string keys;
        std::string values;
        for (auto it = data.begin(); it != data.end(); ++it) {
            if (it != data.begin()) {
                keys += ", ";
                values += ", ";
            }
            keys += it->first;
            values += "'" + it->second + "'";
        }
        return "INSERT INTO " + table + " (" + keys + ") VALUES (" + values + ")";
    }

    static std::string delete_(const std::string& table, 
                              const std::map<std::string, std::string>& where = {}) {
        std::string query = "DELETE FROM " + table;
        if (!where.empty()) {
            query += " WHERE ";
            for (auto it = where.begin(); it != where.end(); ++it) {
                if (it != where.begin()) {
                    query += " AND ";
                }
                query += it->first + "='" + it->second + "'";
            }
        }
        return query;
    }

    static std::string update(const std::string& table, 
                              const std::map<std::string, std::string>& data, 
                              const std::map<std::string, std::string>& where = {}) {
        if (data.empty()) {
            return "UPDATE " + table + " SET ";
        }
        std::string update_str;
        for (auto it = data.begin(); it != data.end(); ++it) {
            if (it != data.begin()) {
                update_str += ", ";
            }
            update_str += it->first + "='" + it->second + "'";
        }

        std::string query = "UPDATE " + table + " SET " + update_str;
        if (!where.empty()) {
            query += " WHERE ";
            for (auto it = where.begin(); it != where.end(); ++it) {
                if (it != where.begin()) {
                    query += " AND ";
                }
                query += it->first + "='" + it->second + "'";
            }
        }
        return query;
    }
};